### PR TITLE
Implement .native-descriptor for async sockets

### PR DESCRIPTION
**Pull request by Kaiepi**
This will allow users to get the native descriptor of the sockets later
on. See https://github.com/perl6/nqp/pull/505 and https://github.com/MoarVM/MoarVM/pull/980

### CODE REVIEW

**Comment by Kaiepi** [comment]
At some point while I was working on this, `S32-io/IO-Socket-Async.t` stopped passing. I'll fix it as soon as possible

**Comment by Kaiepi** [comment]
This now passes all tests

**Comment by Kaiepi** [comment]
Could I get a review on this and the other pullreqs now that the work is done?

File: src/core/IO/Socket/Async.pm6 [hunk 1]
```diff
@@ -242,42 +267,44 @@ my class IO::Socket::Async {
...
+                }, :$VMIO, :$socket-host, :$socket-port;
                 tap($tap);
+
                 CATCH {
                     default {
-                        tap($tap = ListenSocket.new({ Nil },
-                            :$socket-host, :$socket-port)) unless $tap;
+                        $tap = ListenSocket.new({ Nil }, :$VMIO,
+                            :$socket-host, :$socket-port) unless $tap;
+                        tap($tap);
```
Comment (ugexe) [inline src/core/IO/Socket/Async.pm6]: I can't tell if this is intended or not, but the logic is not quite the same anymore.
Before:
```
unless $tap {
    tap($tap = ...)
}
```
After:
```
unless $tap {
    $tap = ...;
}
tap($tap)
```

Comment (Kaiepi) [reply src/core/IO/Socket/Async.pm6]: Unintended

File: src/core/IO/Socket/Async.pm6 [hunk 2]
```diff
@@ -242,42 +267,43 @@ my class IO::Socket::Async {
...
                     $p
-                }, :$socket-host, :$socket-port;
+                }, :$VMIO, :$socket-host, :$socket-port;
                 tap($tap);
+
                 CATCH {
                     default {
-                        tap($tap = ListenSocket.new({ Nil },
-                            :$socket-host, :$socket-port)) unless $tap;
+                        tap(ListenSocket.new: { Nil }, :$VMIO,
```
Comment (ugexe) [inline src/core/IO/Socket/Async.pm6]: This is still not equivalent to the previous code, which assigned the `ListenSocket` to `$tap`

Comment (Kaiepi) [reply src/core/IO/Socket/Async.pm6]: Oop

File: src/core/IO/Socket/Async.pm6 [hunk 3]
```diff
@@ -165,42 +166,65 @@ my class IO::Socket::Async {
...
     class ListenSocket is Tap {
+        has         $!VMIO;
         has Promise $.socket-host;
         has Promise $.socket-port;
 
-        method new(&on-close, Promise :$socket-host, Promise :$socket-port) {
-            self.bless(:&on-close, :$socket-host, :$socket-port);
+        submethod BUILD(Mu :$!VMIO, Promise :$!socket-host, Promise :$!socket-port) { }
+
+        method new(&on-close, Mu :$VMIO, Promise :$socket-host, Promise :$socket-port) {
```
Comment (ugexe) [inline src/core/IO/Socket/Async.pm6]: Should `$VMIO be required` ala `Mu :$VMIO!`?

Comment (Kaiepi) [reply src/core/IO/Socket/Async.pm6]: I don't see why not

**Comment by zoffixznet** [comment]
👍 Thanks!

**Pull request merged**